agent/consul/state/catalog_events.go

```diff
@@ -0,0 +1,475 @@
+				// our mechanism for connect events is to convert them so we generate
+				// the regular one, convert it to Connect topic and then discard the
+				// original.
+				e := newServiceHealthEventDeregister(changes.Index, before)
+				events = append(events, serviceHealthToConnectEvents(e)...)
+			}
+		}
+
+		if _, ok := nodeChanges[tuple.Node]; ok {
+			// We already rebuilt events for everything on this node, no need to send
+			// a duplicate.
+			continue
+		}
+		// Build service event and append it
+		e, err := newServiceHealthEventForService(tx, changes.Index, tuple)
+		if err != nil {
+			return nil, err
+		}
+		events = append(events, e)
+	}
+
+	// Duplicate any events that affected connect-enabled instances (proxies or
+	// native apps) to the relevant Connect topic.
+	events = append(events, serviceHealthToConnectEvents(events...)...)
+
+	return events, nil
+}
+
+type changeType uint8
+
+const (
+	// changeIndirect indicates some other object changed which has implications
+	// for the target object.
+	changeIndirect changeType = iota
+	changeDelete
+	changeCreate
+	changeUpdate
+)
+
+func changeTypeFromChange(change memdb.Change) changeType {
+	switch {
+	case change.Deleted():
+		return changeDelete
+	case change.Created():
+		return changeCreate
+	default:
+		return changeUpdate
+	}
+}
+
+// serviceHealthToConnectEvents converts already formatted service health
+// registration events into the ones needed to publish to the Connect topic.
+// This essentially means filtering out any instances that are not Connect
+// enabled and so of no interest to those subscribers but also involves
+// switching connection details to be the proxy instead of the actual instance
+// in case of a sidecar.
+func serviceHealthToConnectEvents(events ...stream.Event) []stream.Event {
+	var serviceHealthConnectEvents []stream.Event
+	for _, event := range events {
+		if event.Topic != TopicServiceHealth {
+			// Skip non-health or any events already emitted to Connect topic
+			continue
+		}
+		node := getPayloadCheckServiceNode(event.Payload)
+		if !isConnectEnabledNodeService(node.Service) {
+			continue
+		}
+
+		connectEvent := event
+		connectEvent.Topic = TopicServiceHealthConnect
+
+		// If this is a proxy, set the key to the destination service name.
+		// TODO(streaming): what key to use for Kind==ServiceKindTerminatingGateway ?
+		if node.Service.Kind == structs.ServiceKindConnectProxy {
+			connectEvent.Key = node.Service.Proxy.DestinationServiceName
+		}
+
+		serviceHealthConnectEvents = append(serviceHealthConnectEvents, connectEvent)
+	}
+
+	return serviceHealthConnectEvents
+}
+
+// TODO(streaming): add test case for Kind==ServiceKindTerminatingGateway
+func isConnectEnabledNodeService(sn *structs.NodeService) bool {
+	if sn == nil {
+		return false
+	}
+	switch sn.Kind {
+	case structs.ServiceKindConnectProxy, structs.ServiceKindTerminatingGateway:
+		return true
+	}
+	return sn.Connect.Native
+}
```

Comment:
@hashicorp/consul-service-mesh does this look correct to you? 

ConnectProxy, TerminatingGateway, and Connect.Native are the "connect enabled" services? 

This is related to: https://github.com/hashicorp/consul/pull/8357#discussion_r466334568 and https://github.com/hashicorp/consul/pull/8357#discussion_r466340140

So I think if the service is returned from `/health/connect/:service` it should return `true` here. I started to trace that code, but it is involved, so I thought it would be safer to ask.

Reply:
Yep, that's right. Terminating gateways are returned by the `/health/connect/` endpoint. 

This looks good to me. Connect proxies, terminating gateways, and Connect native apps are the three possibilities.